src/luarocks/core/cfg.lua

```diff
@@ -399,8 +399,8 @@ local function make_defaults(lua_version, target_cpu, platforms, home)
       defaults.lib_extension = "so" -- can be overridden in the config file for mingw builds
       defaults.arch = "cygwin-"..target_cpu
       defaults.cmake_generator = "Unix Makefiles"
-      defaults.variables.CC = "echo -llua | xargs gcc"
-      defaults.variables.LD = "echo -llua | xargs gcc"
+      defaults.variables.CC = os.getenv("CC") or "echo -llua | xargs gcc"
+      defaults.variables.LD = os.getenv("LD") or "echo -llua | xargs gcc"
```

Comment:
This one is a bit weird

Reply:
Seriously, not that this PR added them, but why `echo -llua | xargs gcc` instead of `gcc -llua`?

Reply:
My guess is that CC and LD are used to build more complex command lines using string concatenation, and the xargs part adds `-llua` at the end of that command line, which is required by some linkers.

Reply:
Which also means that this is probably incorrect, because you can replace the `gcc` with `os.getenv("CC")`, but you still need the `echo` and `xargs` stuff.

Reply:
I don't see how that's possible. The `echo foo | xargs bar` construct would not allow for being appended to anything else (it would supplant it and kill off any prior piping), and anything else being append to this construct could just as easily be appended to `bar`. Searching the source code in fact I don't see any usage where a value of just `gcc` wouldn't function the same. It is just concatenating `CC .. " " .. CFLAGS`.

Reply:
I assume that LuaRocks does something like `os.execute(CC .. " " .. options_and_files)` but cygwin requires `os.execute(CC .. " " .. options_and_files .. " -llua")`. This is what the `echo` and `xargs` stuff is supposed to facilitate.

Reply:
Oh interesting, I do see now. Thanks for the explanation. That's probably to tricky a stunt to pull without a comment about why it is that way, but yes it does get that job done.

```console
$ echo foo | xargs echo bar baz
bar baz foo
```

Reply:
How about:

```lua
defaults.variables.CC = "echo -llua | xargs " .. (os.getenv("CC") or "gcc")
```

Reply:
What if the `CC` being supplied is `clang` or `cc` or something else that doesn't accept the `-llua` argument or wants it in a different order?

Reply:
The linking requirement of the Lua (shared) library is actually a Windows thing. Windows doesn't support exporting symbols from the executable to shared extension libraries like Linux does. The `-l` flag is pretty standard too. The lookup order of symbols in libraries/objects on the command line is specified by POSIX -- some linkers are order-independent as an extension. `clang` and `gcc` (including MinGW) should be fine, and I don't know what other C compilers we are likely to encounter in a cygwin environment.

Reply:
And you can still modify `variables.CC` in your LuaRocks config.

Reply:
I like this: 
```
defaults.variables.CC = "echo -llua | xargs " .. (os.getenv("CC") or "gcc")
```

Reply:
> I like this:
> 
> ```
> defaults.variables.CC = "echo -llua | xargs " .. (os.getenv("CC") or "gcc")
> ```

applied thanks